usaspending_api/common/tests/unit/test_tracing.py

```diff
@@ -40,171 +82,185 @@ def caplog(caplog):
     when wrapped in the SubprocessTracer"""
 
     # Enable log output for this logger for duration of this test
-    caplog.set_level(logging.DEBUG, DatadogLoggingTraceFilter._log.name)
+    caplog.set_level(logging.INFO)
     test = f"{inspect.stack()[0][3]}"
+
     # And also send its output through a multiprocessing queue to surface logs from the subprocess
     log_queue = mp.Queue()
-    DatadogLoggingTraceFilter._log.addHandler(QueueHandler(log_queue))
-    DatadogLoggingTraceFilter.activate()
 
-    subproc_test_msg = f"a test message was logged in a subprocess of {test}"
-    state = mp.Queue()
-    stop_sentinel = "-->STOP<--"
+    with tracer.start_as_current_span(name=f"{test}_operation", kind=SpanKind.INTERNAL) as span:
+        span_attributes = {"service.name": f"{test}_service", "resource.name": f"{test}_resource", "span.type": "TEST"}
+        for k, v in span_attributes.items():
+            span.set_attribute(k, v)
 
-    with ddtrace.tracer.trace(
-        name=f"{test}_operation",
-        service=f"{test}_service",
-        resource=f"{test}_resource",
-        span_type=SpanTypes.TEST,
-    ) as span:
-        trace_id = span.trace_id
-        logger = logging.getLogger(f"{test}_logger")
+        trace_id = span.get_span_context().trace_id
+        span_id = span.get_span_context().span_id
         test_msg = f"a test message was logged during {test}"
         logger.warning(test_msg)
+
+        logger.info(f"The corresponding trace id: {trace_id}")
+        logger.info(f"The corresponding span id: {span_id}")
+
+        # Create and start a subprocess, passing the queue to capture logs
         ctx = mp.get_context("fork")
         worker = ctx.Process(
             name=f"{test}_subproc",
             target=_do_things_in_subproc,
-            args=(
-                subproc_test_msg,
-                state,
-            ),
+            args=(caplog, log_queue, trace_id),  # Pass the main trace ID to the subprocess
         )
         worker.start()
-        worker.join(timeout=10)
-        DatadogLoggingTraceFilter._log.warning(stop_sentinel)
+        worker.join(timeout=100)
+
+        if worker.is_alive():
+            worker.terminate()
+            try:
+                log_records = []
+                draining = True
+                while draining:
+                    while not log_queue.empty():
+                        log_record = log_queue.get(block=True, timeout=5)
+                        log_records.append(log_record)
+
+                    if log_queue.empty():
+                        draining = False
+
+                for log_record in log_records:
+                    caplog.handler.flush(log_record)
+            except Exception:
+                print("Error draining captured log queue when handling subproc TimeoutError")
+                pass
+            raise mp.TimeoutError(f"subprocess {worker.name} did not complete in timeout")
```

Comment:
Why does it raise an error regardless after the `try...except`?

Reply:
I think there are a few things going on here. The `if worker.is_alive()` block handles the case where the subprocess did not complete within the specified timeout period `worker.join(timeout=100)`. In this case, we ultimately want to fail the test, but we first want to drain the logs from our `log_queue` (a `multiprocessing.Queue`) to preserve that debugging information and potentially figure out why it missed the timeout window. After draining the logs, we're safe to go ahead and throw a meaningful exception to fail the test.